### PR TITLE
sensors: lis2dh: fix runtime and 16g range compile-time selectable range conversion

### PR DESCRIPTION
**Pull request by mniestroj**
This is an alternative implementation of PR #19983

### CODE REVIEW

**Comment by avisconti** [comment]
This PR is addressing #19872

**Comment by avisconti** [comment]
I'm ok withis changes, but we need to address the following:

As you probably know we are using the same driver for different devices that has same registers interface (i.e. whoami is same: 0x33). This devices currently are: LIS2DH, LIS3DH, LSM303DLHC, LIS2DH12, LSM303AGR (see drivers/sensor/lis2dh/Kconfig).

Now the issue is that, even if reg i/f is same, they have different mechanical sensors, so the scale is different. I was using the lsm303agr ones...

**Comment by avisconti** [comment]
> I'm ok withis changes, but we need to address the following:
> 
> As you probably know we are using the same driver for different devices that has same registers interface (i.e. whoami is same: 0x33). This devices currently are: LIS2DH, LIS3DH, LSM303DLHC, LIS2DH12, LSM303AGR (see drivers/sensor/lis2dh/Kconfig).
> 
> Now the issue is that, even if reg i/f is same, they have different mechanical sensors, so the scale is different. I was using the lsm303agr ones...

As @mniestroj  noted, seems that only LSM303AGR has a different set of gains, so we need to implement this in the driver.

**Review by avisconti — CHANGES_REQUESTED** [review]
Please address what discussed previously, i.e. addtressing the LSM303AGR set of gains.

**Comment by mniestroj** [comment]
@avisconti 
I have just rebased this MR on top of master, in order to get rid of conflicts.

@sdalu 
Could you test if this MR solves #19872 ?

File: drivers/sensor/lis2dh/lis2dh.c [hunk 1]
```diff
@@ -13,6 +13,26 @@
 LOG_MODULE_REGISTER(lis2dh, CONFIG_SENSOR_LOG_LEVEL);
 #include "lis2dh.h"
 
+#if defined(DT_INST_0_ST_LSM303AGR_ACCEL)
+/*
+ * Use values for 16g low-power mode in LSM303AGR DS "Table 3. Sensor
+ * characteristics" and LIS2DH DS "Table 3. Mechanical characteristics".
+ */
+#define ACCEL_SCALE_FACTOR(x)		((x) * 18758LL / 19200LL)
```
Comment (avisconti) [inline drivers/sensor/lis2dh/lis2dh.c]: So you decided to use a matematic expression.
Actually I preferred to have a separate set of scales for the two different sensors.
I think it is easier to find the correspondent values on the datasheets.
What do you think?

**Review by avisconti — CHANGES_REQUESTED** [review]

File: drivers/sensor/lis2dh/lis2dh.c (hunk 1 above)
Comment (mniestroj) [reply drivers/sensor/lis2dh/lis2dh.c]:
```
#define ACCEL_SCALE(sensitivity)		\
	(SENSOR_G * (sensitivity) >> 14 / 100)

/*
 * Use values for low-power mode in DS "Mechanical (Sensor) characteristics",
 * multiplied by 100.
 */
static const u32_t lis2dh_reg_val_to_scale[] = {
#if defined(DT_INST_0_ST_LSM303AGR_ACCEL)
	ACCEL_SCALE(1563),
	ACCEL_SCALE(3126),
	ACCEL_SCALE(6252),
	ACCEL_SCALE(18758),
#else
	ACCEL_SCALE(1600),
	ACCEL_SCALE(3200),
	ACCEL_SCALE(6400),
	ACCEL_SCALE(19200),
#endif
};
```

Something like that? :)

Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh.c]: > ```
> #define ACCEL_SCALE(sensitivity)		\
> 	(SENSOR_G * (sensitivity) >> 14 / 100)
> 
> /*
>  * Use values for low-power mode in DS "Mechanical (Sensor) characteristics",
>  * multiplied by 100.
>  */
> static const u32_t lis2dh_reg_val_to_scale[] = {
> #if defined(DT_INST_0_ST_LSM303AGR_ACCEL)
> 	ACCEL_SCALE(1563),
> 	ACCEL_SCALE(3126),
> 	ACCEL_SCALE(6252),
> 	ACCEL_SCALE(18758),
> #else
> 	ACCEL_SCALE(1600),
> 	ACCEL_SCALE(3200),
> 	ACCEL_SCALE(6400),
> 	ACCEL_SCALE(19200),
> #endif
> };
> ```
> 
> 
> Something like that? :)

Yes, exactly! :)

**Comment by sdalu** [comment]
@mniestroj just tested, got what seems correct values, it solves  #19872

File: drivers/sensor/lis2dh/lis2dh.c (hunk 1 above)
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh.c]: @mniestroj 
So, are you changing it in this way?

Comment (mniestroj) [reply drivers/sensor/lis2dh/lis2dh.c]: @avisconti Yes, just wanted to test with my board before submission :) It is ready now.

**Review by avisconti — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**